nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/StandardOidcIdentityProvider.java

```diff
@@ -0,0 +1,339 @@
+            final String rawClientId = properties.getOidcClientId();
+            if (StringUtils.isBlank(rawClientId)) {
+                throw new RuntimeException("Client ID is required when configuring an OIDC Provider.");
+            }
+            clientId = new ClientID(rawClientId);
+
+            // client secret
+            final String rawClientSecret = properties.getOidcClientSecret();
+            if (StringUtils.isBlank(rawClientSecret)) {
+                throw new RuntimeException("Client secret is required when configured an OIDC Provider.");
+            }
+            clientSecret = new Secret(rawClientSecret);
+
+            try {
+                // retrieve the oidc provider metadata
+                oidcProviderMetadata = retrieveOidcProviderMetadata(properties.getOidcDiscoveryUrl());
+            } catch (IOException | ParseException e) {
+                throw new RuntimeException("Unable to retrieve OpenId Connect Provider metadata from: " + properties.getOidcDiscoveryUrl(), e);
+            }
+
+            // ensure the oidc provider supports basic or post client auth
+            final List<ClientAuthenticationMethod> clientAuthenticationMethods = oidcProviderMetadata.getTokenEndpointAuthMethods();
+            if (clientAuthenticationMethods == null
+                    || (!clientAuthenticationMethods.contains(ClientAuthenticationMethod.CLIENT_SECRET_BASIC)
+                    && !clientAuthenticationMethods.contains(ClientAuthenticationMethod.CLIENT_SECRET_POST))) {
+
+                throw new RuntimeException(String.format("OpenId Connect Provider does not support %s or %s",
+                        ClientAuthenticationMethod.CLIENT_SECRET_BASIC.getValue(),
+                        ClientAuthenticationMethod.CLIENT_SECRET_POST.getValue()));
+            }
+
+            // extract the supported json web signature algorithms
+            final List<JWSAlgorithm> allowedAlgorithms = oidcProviderMetadata.getIDTokenJWSAlgs();
+            if (allowedAlgorithms == null || allowedAlgorithms.isEmpty()) {
+                throw new RuntimeException("The OpenId Connect Provider does not support any JWS algorithms.");
+            }
+
+            try {
+                // get the preferred json web signature algorithm
+                final String rawPreferredJwsAlgorithm = properties.getOidcPreferredJwsAlgorithm();
+
+                final JWSAlgorithm preferredJwsAlgorithm;
+                if (StringUtils.isBlank(rawPreferredJwsAlgorithm)) {
+                    preferredJwsAlgorithm = oidcProviderMetadata.getIDTokenJWSAlgs().get(0);
+                } else {
+                    if ("none".equals(rawPreferredJwsAlgorithm)) {
+                        preferredJwsAlgorithm = null;
+                    } else {
+                        preferredJwsAlgorithm = JWSAlgorithm.parse(rawPreferredJwsAlgorithm);
+                    }
+                }
+
+                if (preferredJwsAlgorithm == null) {
```

Comment:
In version 5.34 of `oauth2-oidc-sdk`, the interface `com.nimbusds.oauth2.sdk.http.ResourceRetriever` is deprecated, as are its implementations.